### ISSUE TITLE
How to change configuration at runtime ?

### ISSUE THREAD

**Issue by ch4mpy** [damienbod/angular-auth-oidc-client#923]
I couldn't find a way to change some configuration after lib initialisation.

My intention is to replace a "default" value for `stsServer` with one taken from a configuration form.

How to achieve this ?

**Comment by FabianGosebrink** [comment]
Hey, currently we do not have an option to change configuration at runtime. You could, however, save the sts server the user gave you in the form, reload the application and provide it through a service. The service reads the sts server the user entered and passes it into the config. Add the servie to the `deps` array then.

**Comment by wolfpackt99** [comment]
Also interested in this. Not to found of browser reload though.

**Comment by damienbod** [comment]
We could probably add support for this, does not seem like a big change or effort

**Comment by ch4mpy** [comment]
Thank you.
Realoding the app is a kind of user experience I prefer to avoid.

**Comment by martijnvanschie** [comment]
Instead of opening a new issue for a related question. I think this is in line with what i'm looking for.
I want to allow different auth servers and the user can select which one (like google, facebook and OpenID).
Is this currently possible.

I used another library in the past and just updated the config when the user selected the IdP.
Regards,

**Comment by damienbod** [comment]
@martijnvanschie This is not supported at present. We plan to support this by using multiple IDP configuration support. We hope it makes version 12.

Greetings Damien

**Comment by martijnvanschie** [comment]
@damienbod that would be a very nice addition. Looking forward to this 😀

**Comment by FabianGosebrink** [comment]
Hey, I just gave this thing a look because we are heavily working on V12.

In V12 we are supporting multiple IDPs. So you can pass multiple configs and of course run with one single config as well.

Here are the things which came into my mind: Technically switching the config is not such a great deal, it is doable. 

If we switch the configuration at runtime, we can not do a logout at the server, because this would cause a redirect to the sts. Which you do not want when you only want to switch configs. So we can only do `logoffLocal` erasing all local login information. That is the first thing which I saw.

Currently, we are _not_ storing the complete config in the storage but only in memory because of performance reasons. We are only storing the `authWellKnownEndpoints`. So when you switch the configuration at runtime we would replace it in the memory, but when you hit refresh in the browser, the original config would be set again. 

Also, when looking from the multiple config perspective we would have to support adding a config, replacing a config and replacing multiple configs with other configs.

And when refreshing all the original configs would be set again. So when replacing the config the user has configured A, but running with B OR has A, B, C, ... configured and is running with D, E and F ... This would lead so inconsistency.

Yeah, that is my two cents on that. I don't know how we can solve this to make it useable for the users and am open for proposals.

Thanks, Fabian

**Comment by CrazyBaran** [comment]
Hi. I struggling right now with such case. We had SaaS application where our tenants user should can setup their own oidc servers. So we right now kept this information on our backend about which sts server they want to use and then send it on app initialization. This is not a case of multiple IDPs cause each tenant can have only one, but how right now load this configuration into angular-auth-oidc-client so we can allow login into app from some sts?

**Comment by FabianGosebrink** [comment]
Hey @CrazyBaran, we support loading the config from an HTTP source. See https://github.com/damienbod/angular-auth-oidc-client/blob/main/projects/sample-code-flow-http-config/src/app/auth-config.module.ts for reference.

**Comment by CrazyBaran** [comment]
@FabianGosebrink Ah nice, so sorry for bother.

**Comment by FabianGosebrink** [comment]
No worries. All good, :-) We are improving the docs in V12 and hope this can be done more obvious then.

**Comment by FabianGosebrink** [comment]
After having no feedback on this, I am afraid we have to postpone this feature. I currently see no way to change the config at runtime without having an inconsistency in the configs.

**Comment by silviu-sterian** [comment]
Hi, I know this thread has been closed, but could you tell me if there is a way to switch between config at runtime at all? I upgraded to v14 and looks like OidcConfigService is not exported anymore or available - so my work around using .withConfigs does not work. I was combining that and overriding storage to load different configs (external providers) in a demo app.
 It looks like I don;t have access anymore to any methods where I can pass in a config and initiate the auth process

**Comment by FabianGosebrink** [comment]
Yes, switching configs at runtime is not possible. Not without a reload.

**Comment by neistow** [comment]
@FabianGosebrink We're heavily blocked by the lack of a runtime configuration change. We're running a multitenant SaaS app where tenant backend URLs are fetched from the api that requires authorization, so we need to update the `secureRoutes` property after we fetch tenant URLs for the built-in interceptor to work.

I've come up with this workaround (or the same concept, but with BehaviorSubject), tested it, and it works. I am wondering - what are the consequences of ~~ab~~using library like this? 

And if there are, maybe a `take(1)` operator should be added in the internal implementation of lib to prevent unintended misuse?

```
@Injectable({ providedIn: 'root' })
export class DynamicConfigurationProvider {
    private _config = signal<OpenIdConfiguration>({
        authority: 'authority',
        redirectUrl: window.location.origin,
        clientId: 'client',
        scope: 'openid profile email',
        responseType: 'code'
    });

    public config = this._config.asReadonly();
    
    public addSecureUrl(url: string): void {
        // Update signal
    }

    public removeSecureUrl(url: string): void {
        // Update signal
    }
}

export const configLoaderFactory = (provider: DynamicConfigurationProvider) => {
    return new StsConfigHttpLoader(toObservable(provider.config));
};

export const authConfig: PassedInitialConfig = {
    loader: {
        provide: StsConfigLoader,
        useFactory: configLoaderFactory,
        deps: [DynamicConfigurationProvider],
    },
};

export const appConfig: ApplicationConfig = {
    providers: [
        provideAuth(authConfig),
        provideHttpClient(
            withInterceptors([authInterceptor()])
        )
    ]
};
```
